File: app/src/main/java/com/geekorum/ttrss/data/ArticlesDatabaseModule.java
/*
 * Geekttrss is a RSS feed reader application on the Android Platform.
 *
 * Copyright (C) 2017-2019 by Frederic-Charles Barthelery.
 *
 * This file is part of Geekttrss.
 *
 * Geekttrss is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Geekttrss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Geekttrss.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.geekorum.ttrss.data;

import android.app.Application;
import androidx.room.Room;
import androidx.sqlite.db.SupportSQLiteOpenHelper;
import com.geekorum.ttrss.data.migrations.MigrationFrom1To2;
import com.geekorum.ttrss.data.migrations.MigrationFrom2To3;
import com.geekorum.ttrss.data.migrations.MigrationFrom3To4;
import com.geekorum.ttrss.data.migrations.MigrationFrom4To5;
import com.geekorum.ttrss.data.migrations.MigrationFrom5To6;
import com.geekorum.ttrss.data.migrations.MigrationFrom6To7;
import com.geekorum.ttrss.providers.ArticlesProvidersDao;
import dagger.Module;
import dagger.Provides;

import javax.inject.Singleton;

/**
 * Provides some dependencies needed to access the room database.
 */
@Module
public class ArticlesDatabaseModule {

    @Provides
    @Singleton
    ArticlesDatabase providesAppDatabase(Application application) {
        return Room.databaseBuilder(application, ArticlesDatabase.class, ArticlesDatabase.DATABASE_NAME)
                .addMigrations(MigrationFrom1To2.INSTANCE,
                        MigrationFrom2To3.INSTANCE,
                        MigrationFrom3To4.INSTANCE,
                        MigrationFrom4To5.INSTANCE,
                        MigrationFrom5To6.INSTANCE,
                        MigrationFrom6To7.INSTANCE)
                .build();
    }

    @Provides
    SupportSQLiteOpenHelper providesRoomDbHelper(ArticlesDatabase database) {
        return database.getOpenHelper();
    }

    @Provides
    ArticleDao providesArticleDao(ArticlesDatabase database) {
        return database.articleDao();
    }

    @Provides
    TransactionsDao providestransactionsDao(ArticlesDatabase database) {
        return database.transactionsDao();
    }

    @Provides
    SynchronizationDao providesSynchronizationDao(ArticlesDatabase database) {
        return database.synchronizationDao();
    }

    @Provides
    ArticlesProvidersDao providesArticlesProvidersDao(ArticlesDatabase database) {
        return database.articlesProvidersDao();
    }

    @Provides
    FeedsDao providesFeedsDao(ArticlesDatabase database) {
        return database.feedsDao();
    }

}

File: app/src/main/java/com/geekorum/ttrss/data/ArticlesDatabaseModule.kt
/*
 * Geekttrss is a RSS feed reader application on the Android Platform.
 *
 * Copyright (C) 2017-2020 by Frederic-Charles Barthelery.
 *
 * This file is part of Geekttrss.
 *
 * Geekttrss is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Geekttrss is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Geekttrss.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.geekorum.ttrss.data

import android.app.Application
import androidx.room.Room
import androidx.sqlite.db.SupportSQLiteOpenHelper
import com.geekorum.ttrss.data.migrations.ALL_MIGRATIONS
import com.geekorum.ttrss.providers.PurgeArticlesDao
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.android.components.ApplicationComponent
import javax.inject.Singleton

/**
 * Provides some dependencies needed to access the room database.
 */
@Module
@InstallIn(ApplicationComponent::class)
object ArticlesDatabaseModule {
    @Provides
    @Singleton
    internal fun providesAppDatabase(application: Application?): ArticlesDatabase {
        return Room.databaseBuilder(application!!, ArticlesDatabase::class.java, ArticlesDatabase.DATABASE_NAME)
                .fallbackToDestructiveMigrationOnDowngrade()
                .addMigrations(*ALL_MIGRATIONS.toTypedArray())
                .build()
    }

    @Provides
    internal fun providesRoomDbHelper(database: ArticlesDatabase): SupportSQLiteOpenHelper {
        return database.openHelper
    }

    @Provides
    internal fun providesArticleDao(database: ArticlesDatabase): ArticleDao {
        return database.articleDao()
    }

    @Provides
    internal fun providesTransactionsDao(database: ArticlesDatabase): TransactionsDao {
        return database.transactionsDao()
    }

    @Provides
    internal fun providesSynchronizationDao(database: ArticlesDatabase): SynchronizationDao {
        return database.synchronizationDao()
    }

    @Provides
    internal fun providesPurgeArticlesDao(database: ArticlesDatabase): PurgeArticlesDao {
        return database.articlesProvidersDao()
    }

    @Provides
    internal fun providesFeedsDao(database: ArticlesDatabase): FeedsDao {
        return database.feedsDao()
    }

    @Provides
    internal fun providesAccountInfoDao(database: ArticlesDatabase): AccountInfoDao {
        return database.accountInfoDao()
    }

    @Provides
    internal fun providesManageFeedsDao(database: ArticlesDatabase) : ManageFeedsDao {
        return database.manageFeedsDao()
    }
}
